bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/events/BuildManager.java

```diff
@@ -740,14 +740,47 @@ public void endTopLevel(boolean needsBuild) {
 		autoBuildJob.build(needsBuild);
 	}
 
+	/**
+	 * Waits till autobuild finished. Tries to finish it as soon as possible.
+	 */
 	public void waitForAutoBuild() {
-		autoBuildJob.wakeUp(0);
-		try {
-			Job.getJobManager().join(ResourcesPlugin.FAMILY_AUTO_BUILD, null);
-		} catch (OperationCanceledException | InterruptedException e) {
-			// ignore
+		while (!(autoBuildJob.getState() == Job.NONE)) {
```

Comment:
The two methods seem to be doing same, the only difference is the job - so it can be one method with argument

Reply:
yea, there is always room for optimization

Reply:
Please merge in one method